### PR TITLE
Save Sdist to disk and use PKG-INFO metadata

### PR DESCRIPTION
**Pull request by tdejager**
Making some progress on: https://github.com/prefix-dev/rip/issues/32

Should now save sdists when running the resolution step.

E.g
```bash
cargo r sdist
cargo r numpy==1.7.0
```

This uses the PKG-INFO metadata https://peps.python.org/pep-0643/ to resolve the metadata for the source distribution. Any [PEP517](https://peps.python.org/pep-0517/#source-distributions) conformant build back-end should make this. So these are a lot. 

The `cargo r apache-airflow[all]` seems to use the sdist resolution as well, although it fails on a package that does not meet my python interpreter requirement, will look into this later if this is the case.

Next up, sdist building, stay tuned!

### CODE REVIEW

**Comment by wolfv** [comment]
Nice! Just a little issue with the conflict.

File: crates/index/src/main.rs [hunk 1]
```diff
@@ -107,7 +107,7 @@ pub async fn index(index_url: Url) -> Result<(), miette::Error> {
 
         // Continue if there was an error in downloading and skip for now :)
         let metadata = if let Some(metadata) = metadata {
-            metadata.1
+            metadata.unwrap().1
```
Comment (wolfv) [inline crates/index/src/main.rs:110]: tbh reading this is a bit confusing :D Is there a way to make this a bit more clear?

File: crates/rattler_installs_packages/src/resolve.rs [hunk 2]
```diff
@@ -167,10 +169,9 @@ impl<'db, 'i> PypiDependencyProvider<'db, 'i> {
         }
 
         // Filter only artifacts we can work with
-        artifacts.retain(|a| a.is::<Wheel>());
         if artifacts.is_empty() {
             // If there are no wheel artifacts, we're just gonna skip it
-            return Err("there are no wheels available");
+            return Err("there are no artifacts available");
```
Comment (wolfv) [inline crates/rattler_installs_packages/src/resolve.rs:174]: Maybe we should talk about `packages` instead? Not sure the average person understands what we mean with an artifact.

File: crates/index/src/main.rs (hunk 1 above)
Comment (wolfv) [reply crates/index/src/main.rs:110]: And maybe if this is a safe unwrap, you could add a little comment? I think Bas would ask for an expect either way :D

Comment (tdejager) [reply crates/index/src/main.rs:110]: Yeah so I'm a bit unsure about this index program file as it's kinda half-baked, would rather pull it out of this project entirely and make it kinda nice. Do you still want met to change it?

File: crates/rattler_installs_packages/src/resolve.rs (hunk 2 above)
Comment (tdejager) [reply crates/rattler_installs_packages/src/resolve.rs:174]: Sure!

**Review by wolfv — APPROVED** [review]

File: crates/rattler_installs_packages/src/sdist.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        // Loop over entries
+        for entry in archive.entries().into_diagnostic()? {
+            let mut entry = entry.into_diagnostic()?;
+
+            // Find PKG-INFO and parse this file
+            if entry.path().into_diagnostic()?.ends_with("PKG-INFO") {
+                let mut bytes = Vec::new();
+                entry.read_to_end(&mut bytes).into_diagnostic()?;
+                let metadata = Self::parse_metadata(&bytes)?;
+                return Ok((bytes, metadata));
```
Comment (pradyunsg) [inline crates/rattler_installs_packages/src/sdist.rs:48]: This should only use this metadata from the PKG-INFO file if the metadata version is 2.2 or greater. Using on older versions means that you'd rely on the metadata not required to be reliable.

**Pull request merged**